File: src/main/java/zwylair/randomitembattle/commands/GameManage.java
package zwylair.randomitembattle.commands;

import java.util.ArrayList;
import java.util.List;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.context.CommandContext;

import net.minecraft.server.MinecraftServer;
import net.minecraft.server.command.CommandManager;
import net.minecraft.server.command.ServerCommandSource;
import static net.minecraft.server.command.CommandManager.literal;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.text.Text;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.GameMode;

import static zwylair.randomitembattle.RandomItemBattle.*;
import static zwylair.randomitembattle.commands.ConfigureWorld.isWorldNotConfigured;
import static zwylair.randomitembattle.utils.PosCalculator.posToCoord;


public class GameManage {
    public static void register(CommandDispatcher<ServerCommandSource> dispatcher) {
        dispatcher.register(literal("rib").then(literal("game").then(literal("start")
                .executes(GameManage::startGame))));
        dispatcher.register(literal("rib").then(literal("game").then(literal("stop")
                .executes(GameManage::stopGame))));

        dispatcher.register(literal("rib").then(literal("items").then(literal("pause")
                .executes(GameManage::pauseItemSpawning))));
        dispatcher.register(literal("rib").then(literal("items").then(literal("resume")
                .executes(GameManage::resumeItemSpawning))));
    }
    
    public static int startGame(CommandContext<ServerCommandSource> ctx) {
        ServerWorld world = ctx.getSource().getWorld();
        List<ServerPlayerEntity> players = world.getPlayers();

        if (isWorldNotConfigured(world)) { ConfigureWorld.configureWorld(ctx); }
        if (centerPosition == null) {
            ctx.getSource().sendFeedback(() -> Text.literal(chatModPrefix + "§cThe center position is not configured! Configure the center position with (/rib set_center_pos)"), false);
            return 1;
        }
        if (playerPositions.isEmpty()) {
            ctx.getSource().sendFeedback(() -> Text.literal(chatModPrefix + "§cThere is no one position to start with!"), false);
            return 1;
        }
        if (world.getPlayers().size() > playerPositions.size()) {
            ctx.getSource().sendFeedback(() -> Text.literal(String.format(chatModPrefix + "§cThere are only %d positions for %d players!", playerPositions.size(), players.size())), false);
            return 1;
        }

        // shortcuts
        Vec3d cPos = centerPosition;
        MinecraftServer server = ctx.getSource().getServer();
        CommandManager commandManager = server.getCommandManager();
        ServerCommandSource commandSource = server.getCommandSource();
        List<Vec3d> playerPositionsLeft = new ArrayList<>(playerPositions);

        // clean and place procedures in separate forEach because the server chaotic
        // executes commands, so we need some time interval before teleporting players

        // clean center and every startpos
        commandManager.executeWithPrefix(commandSource, String.format("/fill %s %s %s %s %s %s air", posToCoord(cPos.x) - 15, (int) cPos.y - 5, posToCoord(cPos.z) - 15, posToCoord(cPos.x) + 15, (int) cPos.y + 20, posToCoord(cPos.z) + 15));
        playerPositionsLeft.forEach((pos) -> commandManager.executeWithPrefix(commandSource, String.format("/fill %s %s %s %s %s %s air", posToCoord(pos.x) - 15, (int) pos.y - 5, posToCoord(pos.z) - 15, posToCoord(pos.x) + 15, (int) pos.y + 20, posToCoord(pos.z) + 15)));

        // set bedrock blocks
        commandManager.executeWithPrefix(commandSource, String.format("/fill %s %s %s %s %s %s bedrock", posToCoord(cPos.x) - 1, (int) cPos.y - 1, posToCoord(cPos.z) - 1, posToCoord(cPos.x) + 1, (int) cPos.y - 1, posToCoord(cPos.z) + 1));
        playerPositionsLeft.forEach((pos) -> commandManager.executeWithPrefix(commandSource, String.format("/setblock %s %s %s bedrock", posToCoord(pos.x), (int) pos.y - 1, posToCoord(pos.z))));

        commandManager.executeWithPrefix(commandSource, "/kill @e[type=!player]");

        // prepare players (also interval)
        players.forEach((player) -> {
            player.getInventory().clear();
            player.changeGameMode(GameMode.SURVIVAL);
            player.getHungerManager().setFoodLevel(20);
            player.heal(20);
        });

        // teleport players to random startpos
        players.forEach((player) -> {
            int randomPositionIndex = world.getRandom().nextInt(playerPositionsLeft.size());
            Vec3d randomPosition = playerPositionsLeft.get(randomPositionIndex);

            player.teleport(randomPosition.x, randomPosition.y, randomPosition.z);
            playerPositionsLeft.remove(randomPositionIndex);
        });

        itemSpawningStatus = true;
        resetWaitedTicks = true;
        isGameStarted = true;
        return 0;
    }

    public static int stopGame(CommandContext<ServerCommandSource> ctx) {
        if (!isGameStarted) {
            ctx.getSource().sendFeedback(() -> Text.literal(chatModPrefix + "§cThe game is not started!"), false);
            return 1;
        }

        ctx.getSource().getWorld().getPlayers().forEach((player) -> {
            player.changeGameMode(GameMode.CREATIVE);
            player.getHungerManager().setFoodLevel(20);
            player.heal(20);
        });

        itemSpawningStatus = false;
        resetWaitedTicks = true;
        isGameStarted = false;
        return 0;
    }

    public static int pauseItemSpawning(CommandContext<ServerCommandSource> ctx) {
        itemSpawningStatus = false;

        ctx.getSource().sendFeedback(() -> Text.literal(chatModPrefix + "Item spawning was paused"), false);
        return 0;
    }

    public static int resumeItemSpawning(CommandContext<ServerCommandSource> ctx) {
        itemSpawningStatus = true;

        ctx.getSource().sendFeedback(() -> Text.literal(chatModPrefix + "Item spawning was resumed"), false);
        return 0;
    }
}

File: src/main/kotlin/zwylair/randomitembattle/commands/GameManage.kt
package zwylair.randomitembattle.commands

import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.context.CommandContext

import net.minecraft.text.Text
import net.minecraft.world.GameMode
import net.minecraft.server.command.CommandManager.literal
import net.minecraft.server.command.ServerCommandSource

import zwylair.randomitembattle.RandomItemBattle
import zwylair.randomitembattle.commands.ConfigureWorld.isWorldNotConfigured
import zwylair.randomitembattle.utils.PosCalculator.Companion.posToCoord

object GameManage {
    fun register(dispatcher: CommandDispatcher<ServerCommandSource>) {
        dispatcher.register(literal("rib").then(literal("game").then(literal("start").executes { startGame(it) })))
        dispatcher.register(literal("rib").then(literal("game").then(literal("stop").executes { stopGame(it) })))
        dispatcher.register(literal("rib").then(literal("items").then(literal("pause").executes { pauseItemSpawning(it) })))
        dispatcher.register(literal("rib").then(literal("items").then(literal("resume").executes { resumeItemSpawning(it) })))
    }

    private fun startGame(ctx: CommandContext<ServerCommandSource>): Int {
        val world = ctx.source.world  // shortcut

        if (isWorldNotConfigured(world)) ConfigureWorld.configureWorld(ctx)

        // centerpos is not set
        try {
            RandomItemBattle.centerPosition
        } catch (e: UninitializedPropertyAccessException) {
            ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} §cThe center position is not configured! Configure the center position with (/rib set_center_pos)") }, false)
            return 1
        }

        // there is no startpos
        if (RandomItemBattle.playerPositions.isEmpty()) {
            ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} §cThere is no one position to start with!") }, false)
            return 1
        }

        // there are too many players for startpos count
        if (world.players.size > RandomItemBattle.playerPositions.size) {
            ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} §cThere are only ${RandomItemBattle.playerPositions.size} positions for ${world.players.size} players!") }, false)
            return 1
        }

        // centerpos is too close to world borders
        if (RandomItemBattle.centerPosition.y < -40 || RandomItemBattle.centerPosition.y > 290) {
            ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} ${RandomItemBattle.CHAT_PREFIX_WARN} The Y coordinate is too close to the world borders. This may cause some problems with clearing the playground. Recommended Y coordinate for creating playground = 64.") }, false)
        }

        val cPos = RandomItemBattle.centerPosition  // shortcut
        val server = ctx.source.server  // shortcut
        val playerPositionsLeft = ArrayList(RandomItemBattle.playerPositions)

        // clean and place procedures in separate forEach because the server chaotic
        // executes commands, so we need some time interval before teleporting players

        // clean center and every startpos
        server.commandManager.executeWithPrefix(server.commandSource, "/fill ${posToCoord(cPos.x) - 15} ${cPos.y.toInt() - 5} ${posToCoord(cPos.z) - 15} ${posToCoord(cPos.x) + 15} ${cPos.y.toInt() + 20} ${posToCoord(cPos.z) + 15} air")
        RandomItemBattle.playerPositions.forEach {
            server.commandManager.executeWithPrefix(server.commandSource, "/fill ${posToCoord(it.x) - 15} ${it.y.toInt() - 5} ${posToCoord(it.z) - 15} ${posToCoord(it.x) + 15} ${it.y.toInt() + 20} ${posToCoord(it.z) + 15} air")
        }

        // set bedrock blocks
        server.commandManager.executeWithPrefix(server.commandSource, "/fill ${posToCoord(cPos.x) - 1} ${cPos.y.toInt() - 1} ${posToCoord(cPos.z) - 1} ${posToCoord(cPos.x) + 1} ${cPos.y.toInt() - 1} ${posToCoord(cPos.z) + 1} bedrock")
        RandomItemBattle.playerPositions.forEach {
            server.commandManager.executeWithPrefix(server.commandSource, "/setblock ${posToCoord(it.x)} ${it.y.toInt() - 1} ${posToCoord(it.z)} bedrock")
        }

        server.commandManager.executeWithPrefix(server.commandSource, "/kill @e[type=!player]")

        // prepare players (also interval)
        world.players.forEach {
            it.inventory.clear()
            it.changeGameMode(GameMode.SURVIVAL)
            it.hungerManager.foodLevel = 20
            it.heal(20.0f)
        }

        // teleport players to random startpos
        world.players.forEach {
            val randomPositionIndex = world.random.nextInt(playerPositionsLeft.size)
            val randomPosition = playerPositionsLeft[randomPositionIndex]

            it.teleport(randomPosition.x, randomPosition.y, randomPosition.z)
            playerPositionsLeft.removeAt(randomPositionIndex)
        }

        RandomItemBattle.itemSpawningStatus = true
        RandomItemBattle.resetWaitedTicks = true
        RandomItemBattle.isGameStarted = true
        return 0
    }

    private fun stopGame(ctx: CommandContext<ServerCommandSource>): Int {
        if (!RandomItemBattle.isGameStarted)
        {
            ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} §cThe game is not started!") }, false)
            return 1
        }

        ctx.source.world.players.forEach {
            it.changeGameMode(GameMode.CREATIVE)
            it.hungerManager.foodLevel = 20
            it.heal(20.0f)
        }

        RandomItemBattle.itemSpawningStatus = false
        RandomItemBattle.resetWaitedTicks = true
        RandomItemBattle.isGameStarted = false
        return 0
    }

    private fun pauseItemSpawning(ctx: CommandContext<ServerCommandSource>): Int {
        RandomItemBattle.itemSpawningStatus = false

        ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} Item spawning was paused") }, false)
        return 0
    }

    private fun resumeItemSpawning(ctx: CommandContext<ServerCommandSource>): Int {
        RandomItemBattle.itemSpawningStatus = true

        ctx.source.sendFeedback({ Text.of("${RandomItemBattle.CHAT_PREFIX_MOD} Item spawning was resumed") }, false)
        return 0
    }
}
